RiotSwiftUI/Modules/Template/SimpleUserProfileExample/ViewModel/TemplateUserProfileViewModel.swift

```diff
@@ -0,0 +1,87 @@
+// 
+// Copyright 2021 New Vector Ltd
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+import SwiftUI
+import Combine
+    
+@available(iOS 14.0, *)
+class TemplateUserProfileViewModel: ObservableObject, TemplateUserProfileViewModelProtocol {
+    
+    // MARK: - Properties
+    
+    // MARK: Private
+    private let userService: TemplateUserProfileServiceProtocol
+    private var cancellables = Set<AnyCancellable>()
+    
+    // MARK: Public
+    @Published private(set) var viewState: TemplateUserProfileViewState
+    
+    var completion: ((TemplateUserProfileViewModelResult) -> Void)?
+    
+    private static func defaultState(userService: TemplateUserProfileServiceProtocol) -> TemplateUserProfileViewState {
+        return TemplateUserProfileViewState(avatar: userService.avatarData, displayName: userService.displayName)
+    }
+    
+    // MARK: - Setup
+    init(userService: TemplateUserProfileServiceProtocol, initialState: TemplateUserProfileViewState? = nil) {
+        self.userService = userService
+        self.viewState = initialState ?? Self.defaultState(userService: userService)
+        
+        userService.presencePublisher
+            .map(TemplateUserProfileStateAction.updatePresence)
+            .receive(on: DispatchQueue.main)
+            .sink(receiveValue: self.dispatch(action:))
+            .store(in: &cancellables)
+    }
+    
+    // MARK: - Public methods
+    func proccess(viewAction: TemplateUserProfileViewAction) {
+        switch viewAction {
+        case .cancel:
+            self.cancel()
+        case .done:
+            self.done()
+        }
+    }
+    
+    // MARK: - Private methods
+    /**
+     Send state actions to mutate the state.
+     */
+    private func dispatch(action: TemplateUserProfileStateAction) {
+        var newState = self.viewState
+        reducer(state: &newState, action: action)
+        self.viewState = newState
+    }
+    
+    /**
+     A redux style reducer, all modifications to state happen here. Recieves a state and a state action and produces a new state.
+     */
```

Comment:
Usually in swift I've see the `inout` style used for reducers as it's a bit neater. the alternative is:

```swift
  private func reducer(state: TemplateUserProfileViewState, action: TemplateUserProfileStateAction) -> TemplateUserProfileViewState {
        var newState = state
        switch action {
        case .updatePresence(let presence):
            newState.presence = presence
        }
    return newState
    }
```

Reply:
Don't think it will make a difference to performance. Swift doesn't do a copy on assignment, it optimises things under the cover and only does a copy when there is a mutation to the underlying value.